src/ROOTNTupleWriter.cc

```diff
@@ -0,0 +1,143 @@
+#include <algorithm>
+
+#include "rootUtils.h"
+
+#include "podio/CollectionBase.h"
+#include "podio/EventStore.h"
+#include "podio/ROOTWriter.h"
+
+// ROOT specifc includes
+#include "TFile.h"
+
+#include "podio/ROOTNTupleWriter.h"
+
+using StoreCollection = std::pair<const std::string&, podio::CollectionBase*>;
+
+
+namespace podio {
+
+  ROOTNTupleWriter::ROOTNTupleWriter(const std::string& filename, EventStore* store) :
+    m_events(nullptr),
+    m_metadata(nullptr),
+    m_runMD(nullptr),
+    m_colMD(nullptr),
+    m_evtMD(nullptr),
+    m_file(new TFile(filename.c_str(),"RECREATE","data file")),
+    m_collectionsToWrite(),
+    m_store(store),
+    m_firstEvent(true),
+    m_ntuple_events(nullptr),
+    m_ntuple_metadata(nullptr),
+    m_ntuple_runMD(nullptr),
+    m_ntuple_colMD(nullptr),
+    m_ntuple_evtMD(nullptr)
+  {
+    m_events = rnt::RNTupleModel::Create();
+    m_metadata = rnt::RNTupleModel::Create();
+    m_runMD = rnt::RNTupleModel::Create();
+    m_colMD = rnt::RNTupleModel::Create();
+    m_evtMD = rnt::RNTupleModel::Create();
+  }
+
+  ROOTNTupleWriter::~ROOTNTupleWriter() {
+    // rnt::RNTupleWriter deletes file
+  }
+
+  bool ROOTNTupleWriter::registerForWrite(const std::string& name) {
+    const podio::CollectionBase* tmp_coll(nullptr);
+    if (!m_store->get(name, tmp_coll)) {
+      std::cerr << "no such collection to write, throw exception." << std::endl;
+      return false;
+    }
+    m_collectionsToWrite.push_back(name);
+    return true;
+  }
+
+
+  void ROOTNTupleWriter::writeEvent() {
+  // todo: we only have to do this for the first event, right?
+  std::vector<StoreCollection> collections;
+  collections.reserve(m_collectionsToWrite.size());
+  for (const auto& name : m_collectionsToWrite) {
+    const podio::CollectionBase* coll;
+    m_store->get(name, coll);
+    collections.emplace_back(name, const_cast<podio::CollectionBase*>(coll));
+    collections.back().second->prepareForWrite();
+  }
+
+  if (m_firstEvent) {
+    createModels(collections);
+    m_firstEvent = false;
+  } 
+
+  m_ntuple_events->Fill();
+}
+
+void ROOTNTupleWriter::createModels(const std::vector<StoreCollection>& collections) {
+  for (auto& [name, coll] : collections) {
+    const auto collBuffers = coll->getBuffers();
+    if (collBuffers.data) {
+      // note: these interfaces may still be simplified on the rntuple side
+      auto collClassName = "vector<" + coll->getDataTypeName() +">";
+      std::cout << collClassName << std::endl;
+      auto field = rnt::Detail::RFieldBase::Create(name,  collClassName).Unwrap();
+      // what I cast buffers.data to  does not seem to matter, but it seems I have to cast it to something?
+      void* vv =  (void*) *static_cast<std::vector<float>**>(collBuffers.data); 
```

Comment:
@jblomer this is the cast I am struggling to understand.

The stack trace when this is not casted:
```
1: vector<ExampleWithFixedWidthIntegersData>
41: Warning in <[ROOT.NTuple] Warning /tmp/ubuntu/spack-stage/spack-stage-root-6.26.02-4uthei6jyfh7oowt6ogsl6pakm57qlq4/spack-src/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
41: Fatal: (typedValue->size() % fItemSize) == 0 violated at line 1021 of `/tmp/ubuntu/spack-stage/spack-stage-root-6.26.02-4uthei6jyfh7oowt6ogsl6pakm57qlq4/spack-src/tree/ntuple/v7/src/RField.cxx'
41: aborting
41: #0  0x00007f9370c3cc3a in wait4 () from /lib/x86_64-linux-gnu/libc.so.6
41: #1  0x00007f9370babf67 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
41: #2  0x00007f937139fc9a in TUnixSystem::Exec (shellcmd=<optimized out>, this=0x156d800) at /tmp/ubuntu/spack-stage/spack-stage-root-6.26.02-4uthei6jyfh7oowt6ogsl6pakm57qlq4/spack-src/core/unix/src/TUnixSystem.cxx:2108
41: #3  TUnixSystem::StackTrace (this=0x156d800) at /tmp/ubuntu/spack-stage/spack-stage-root-6.26.02-4uthei6jyfh7oowt6ogsl6pakm57qlq4/spack-src/core/unix/src/TUnixSystem.cxx:2399
41: #4  0x00007f937127d0f2 in DefaultErrorHandler (level=<optimized out>, abort_bool=<optimized out>, location=<optimized out>, msg=0x67ba800 "(typedValue->size() % fItemSize) == 0 violated at line 1021 of `/tmp/ubuntu/spack-stage/spack-stage-root-6.26.02-4uthei6jyfh7oowt6ogsl6pakm57qlq4/spack-src/tree/ntuple/v7/src/RField.cxx'") at /tmp/ubuntu/spack-stage/spack-stage-root-6.26.02-4uthei6jyfh7oowt6ogsl6pakm57qlq4/spack-src/core/base/src/TErrorDefaultHandler.cxx:174
41: #5  0x00007f937132f878 in ErrorHandler(Int_t, const char *, const char *, typedef __va_list_tag __va_list_tag *) (level=6000, location=0x7f9371c2df4c "", fmt=<optimized out>, ap=<optimized out>) at /tmp/ubuntu/spack-stage/spack-stage-root-6.26.02-4uthei6jyfh7oowt6ogsl6pakm57qlq4/spack-src/core/foundation/src/TError.cxx:152
41: #6  0x00007f9371330129 in Fatal (location=location
41: entry=0x7f9371c2df4c "", fmt=<optimized out>) at /tmp/ubuntu/spack-stage/spack-stage-root-6.26.02-4uthei6jyfh7oowt6ogsl6pakm57qlq4/spack-src/core/foundation/src/TError.cxx:249
41: #7  0x00007f9371bce68b in ROOT::Experimental::RVectorField::AppendImpl (this=0x2447860, value=...) at /tmp/ubuntu/spack-stage/spack-stage-root-6.26.02-4uthei6jyfh7oowt6ogsl6pakm57qlq4/spack-src/tree/ntuple/v7/src/RField.cxx:1021
41: #8  0x00007f93720b7e39 in ROOT::Experimental::Detail::RFieldBase::Append (this=0x7391, value=...) at /cvmfs/sw.hsf.org/spackages5/root/6.26.02/x86_64-ubuntu20.04-gcc9.4.0-opt/4uthe/include/ROOT/RField.hxx:203
41: #9  ROOT::Experimental::RNTupleWriter::Fill (this=0x24ae5c0, entry=...) at /cvmfs/sw.hsf.org/spackages5/root/6.26.02/x86_64-ubuntu20.04-gcc9.4.0-opt/4uthe/include/ROOT/RNTuple.hxx:397
41: #10 ROOT::Experimental::RNTupleWriter::Fill (this=0x24ae5c0) at /cvmfs/sw.hsf.org/spackages5/root/6.26.02/x86_64-ubuntu20.04-gcc9.4.0-opt/4uthe/include/ROOT/RNTuple.hxx:392
41: #11 podio::ROOTNTupleWriter::writeEvent (this=<optimized out>) at /home/vavolkl/r/podio/src/ROOTNTupleWriter.cc:73
41: #12 0x0000000000409555 in write<podio::ROOTNTupleWriter> (store=..., writer=...) at /home/vavolkl/r/podio/tests/write_test.h:351
41: #13 0x0000000000405bbd in main () at /home/vavolkl/r/podio/tests/write_rntuple.cpp:24
```

Reply:
This is how the `collBuffers.data` is populated:
https://github.com/AIDASoft/podio/blob/9d078f82d366e9363cbf2e941e3647badbcba690/python/templates/CollectionData.cc.jinja2#L76-L80

The `m_data` member in this case is a `std::unique_ptr<std::vector<XYZData>>` where `XYZ` is a datatype defined in the yaml file and the `Data` class is the corresponding POD that is generated for it.

For reference:
https://github.com/AIDASoft/podio/blob/9d078f82d366e9363cbf2e941e3647badbcba690/python/templates/CollectionData.h.jinja2#L92
https://github.com/AIDASoft/podio/blob/9d078f82d366e9363cbf2e941e3647badbcba690/python/templates/CollectionData.h.jinja2#L24

Reply:
I think this may be a bug of an older RNTuple version.  If I'm not mistaken, with the latest master branch the code should not compile anymore.

The sequence of calls with the latest master has to be like this:
- `model->AddField(std::move(field)) // (for all fields)`
- `model->Freeze(); // no more fields can be added, the default entry becomes available`
- `model->GetDefaultEntry()->CaptureValueUnsafe("fieldName", ptrToObject); // for all fields`

Reply:
To give a bit more context: the idea behind the API change is to clearly separate the two phases of the model life time: phase 1 where the model is built; once built, phase 2 where it can be used to generate entries. We now forbid generating entries in phase 1 to avoid "half-baked entries".